jobhub.go

```diff
@@ -0,0 +1,143 @@
+package jobhub
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"syscall"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+var nextPipelineID int = 1
+
+type Pipeline struct {
+	Name            string
+	Log             logrus.FieldLogger
+	id, nextJobID   int
+	jobContainer    []Job
+	jobByID         map[int]Job
+	jobDependency   map[int][]int
+	isStartingJob   map[int]bool
+	recursionLevels map[int]int
+}
+
+type Job struct {
+	Name, Path     string
+	pipelineID, id int
+}
+
+type exitStatus struct {
+	runtime time.Duration
+	status  syscall.WaitStatus
+}
+
+func NewPipeline() *Pipeline {
+
+	return &Pipeline{
+		Log:             logrus.StandardLogger(),
+		jobByID:         make(map[int]Job),
+		jobDependency:   make(map[int][]int),
+		isStartingJob:   make(map[int]bool),
+		recursionLevels: make(map[int]int),
+	}
+}
+
+func nextIDPipeline() int {
+	tempID := nextPipelineID
+	nextPipelineID++
+	return tempID
+}
+
+func (p *Pipeline) runJob(job Job) (*exitStatus, error) {
+	_, err := os.Stat(job.Path)
+	if err != nil {
+		return nil, err
+	}
+	process := exec.Command(job.Path)
+	start := time.Now()
+	err = process.Run()
+	elapsed := time.Since(start)
+	if err != nil {
+		exitError, ok := err.(*exec.ExitError)
+		if !ok {
+			p.Log.Panicf("Cannot cast to exitError", err)
+		}
+		return &exitStatus{runtime: elapsed, status: exitError.Sys().(syscall.WaitStatus)}, err
+	}
+	return nil, err
+}
+
+func (p *Pipeline) nextIDJob() int {
+	p.nextJobID++
+	return p.nextJobID
+}
+
+func (p *Pipeline) AddJob(job Job) Job {
+	for _, j := range p.jobContainer {
+		if j.id == job.id {
+			p.Log.Panicf("Pipeline [%d][%s] | Job [%d][%s] | Panic: Job has already been added",
+				p.id, p.Name, job.id, job.Name)
+		}
+	}
+	if p.id == 0 {
+		p.id = nextIDPipeline()
+	}
+	job.pipelineID = p.id
+	job.id = p.nextIDJob()
+	p.jobContainer = append(p.jobContainer, job)
+	p.jobByID[job.id] = job
+	p.isStartingJob[job.id] = true
+	return job
+}
+
+func (p *Pipeline) AddJobDependency(job Job, deps ...Job) {
+	for _, d := range deps {
```

Comment:
Also this could be easily tested, before implementing it, write a test which should fail.